### ISSUE TITLE
Auto enqueue of tasks

### ISSUE THREAD

**Issue by tomeramit** [allegroai/trains#95]
For now I use Task.init in the code and then duplicate it on the server and then enqueue.

It is possible to do it from the code? (I mean to send a task to the server with queue number and then automatically the server will send it to the queue)

**Comment by bmartinn** [comment]
Hi @tomeramit ,

You can absolutely do that - use `Task.clone()` to clone an existing task, and `Task.enqueue()` to enqueue the task you've cloned into a queue of your choice, just like you would do using the UI.
Check out the [AutoML example](https://github.com/allegroai/trains/blob/master/examples/automl/task_piping_example.py) where you can see how to use these methods.

**Comment by majdzr** [comment]
Hi @tomeramit, 

A followup question: how do I enqueue a new task, without cloning or duplicating or interacting with the UI? I just want to manually run/execute from cmd new tasks and want them to be waiting "in line" until the current running task is over or I execute them manually (from UI).

Thanks a lot in advance.

**Comment by bmartinn** [comment]
Hi @majdzr,

In order to enqueue an experiment, the experiment must be previously created in the system. In order to create the experiment, you'll need to run the experiment code locally. Afterwards, you can use the newly created experiment ID to programmatically clone it and enqueue it (as seen in the example shared above) - all of this can be done with no interaction with UI.

Does that answer your question? If not, can you please share the differences between this flow and the flow you're trying to create?

**Comment by majdzr** [comment]
@bmartinn thanks for your reply.
What I have in mind is slightly different. Your suggestion requires Base experiment #1 (let's call it B1) to be created first and only then the variants of B1 could be enqueued. 
What I need is having different "bases" to be enqueued, B2 or Bn which are not necessary variants of B1. Today, I use cmd to run B1, wait until finished, run B2, wait etc (on a single GPU), and U was thinking of a good way to manage the queue.

Thanks in advance!

**Comment by bmartinn** [comment]
@majdzr it seems like this is a pipeline processing you are looking for.
If this is the case, there are two options:

1) When each experiment ends, it should enqueue the next experiment in the processing pipeline.
This means that every experiment is responsible for knowing what is the next one. In your example, B1 should know how to clone B2 (template) into B2' and then enqueue B2'.
In order to be clear here, the assumption is that someone had already executed the base experiments B1..Bn , so that the only thing left, is to clone them and enqueue the cloned experiments.
2) An external script (or Jupyter Notebook) orchestrating this entire pipeline, polling the status of experiment B1' then when it is completed, cloning B2 into B2' and enqueue B2' and repeating this process until the pipeline is completed.

BTW: We will soon have an object implementing option (2) as part of  the *Trains* pipeline and automation module.

Is this what you had in mind?

**Comment by majdzr** [comment]
@bmartinn, Thanks for your reply. 
Option (2) is similar to what I had in mind. Will it be part of` _trains-agent_?

For the meantime, what do you suggest to do if while waiting to exp. A to finish (in N hours), I want to enqueue exp. B (that should start only after A is done/failed and/or the GPU is free) (A and B are not necessarily related)?
Thanks in advance.

**Comment by bmartinn** [comment]
Hi @majdzr,

> Option (2) is similar to what I had in mind. Will it be part of` trains-agent?

Option #2 will be part of the _Trains_ platform - this feature is planned to be supported by the both _Trains-Server_ and _Trains-Agent_.

> For the meantime, what do you suggest to do if while waiting to exp. A to finish (in N hours), I want to enqueue exp. B (that should start only after A is done/failed and/or the GPU is free) (A and B are not necessarily related)?

You can use the `APIClient` from _Trains-Agent_ and write a simple monitoring code to wait for experiment A to finish and enqueue experiment B. This script can run on any machine with _Trains-Agent_ installed, even in Jupyter Notebook. An example for that script is:
```python
from time import sleep
from trains_agent import APIClient

minutes_interval = 5

queue_name = "my-queue"
experiment_A = "<experiment-A-id-copied-from-the-UI>"
experiment_B = "<experiment-B-id-copied-from-the-UI>"

queue_id = client.queues.get_all(name=queue_name)[0].id

print("Enqueueing experiment A into queue {}".format(queue_name))
client.tasks.enqueue(queue=queue_id, task=experiment_A)

while True:
  sleep(minutes_interval * 60)

  experiment_A_status = client.tasks.get_by_id(experiment_A).status

  if experiment_A_status == "completed":
    print("Experiment A completed, enqueueing experiment B into queue {}".format(queue_name))
    client.tasks.enqueue(queue=queue_id, task=experiment_B)
    break 
  elif experiment_A_status in ("stopped", "failed"):
    print("Experiment A has stopped running but was not completed (status {}), aborting".format(experiment_A_status))
    break

  print("Experiment A is still running, checking again in {} minutes".format(minutes_interval)

print("Monitor script done")
```

This example script will basically enqueue experiment A, and than check its status every 5 minutes. If the experiment failed to run, the script will stop. If the experiment has completed, the script will enqueue experiment B.

**Comment by H4dr1en** [comment]
> You can use the APIClient from Trains-Agent and write a simple monitoring code to wait for experiment A to finish and enqueue experiment B. This script can run on any machine with Trains-Agent installed, even in Jupyter Notebook.

Is it still the case? Or do workers pull experiments to run from the queues they are listening to automatically?

**Comment by bmartinn** [comment]
Hi @H4dr1en,

The response you quoted was only meant for cases where the user would only want to enqueue experiment B after experiment A was done, to make sure no worker will pull experiment B from the queue and start running it while experiment A is still running (which could happen if you have more than one worker listening on the same queue).

As for your question - you're right - non-busy workers automatically pull experiments from the queue(s) they're listening to.
